### ISSUE TITLE
Comparison operators are now broken when comparing single value with array

### ISSUE THREAD

**Issue by LeaVerou** [mavoweb/mavo#471]
Regression after PR #467 

Testcase: https://codepen.io/leaverou/pen/ZwxrvE?&editors=1100#0
Before merging, the count was 2, now it's 0.

Confirmed that the bug disappears if I revert the changes the PR introduced, and re-appears if I don't.

ping @betaveros . I'm also investigating.
Credits to @karger for discovering the regression.

**Comment by LeaVerou** [comment]
Hi @betaveros I had to partially revert your PR because this was a pretty serious regression and @karger's app was hitting it (and it's an actual app used by a lot of MIT faculty right now). I adjusted the testcase to reproduce both bugs, so we can be sure that both are fixed, and I'll eventually add them as unit tests somewhere.

Could you please look into it as soon as you can, so we can hopefully release 0.2.0 with neither of these bugs? Thanks!

**Comment by betaveros** [comment]
Oops, the bug causing regression was that after deleting the `default`, the comparison operators were interpreted by `addBinaryOperator` as having 0 as the default, which meant they were numeric operators, which meant `$u.numbers` was called on each operand.

On the other hand based on my understanding when I wrote #467, `[count(list("a") = list("a", "b", "a"))]` being 3 is expected behavior and indicates that #404 is actually fixed, although it's evidently not intuitive in the case. I think the rationale is along the lines of: if you write something like `[1, 2, 3] < [4, 5, 6] < [7, 8]` then you'd expect the result to correspond to the expressions `[1 < 4 < 7, 2 < 5 < 8, 3 < 6]`, so `3 < 6 < undefined` must be true i.e. comparisons with elements of an array past its bounds should be true. So the current code, after the partial revert, actually matches my previous expectations in both test cases. Having said that I do agree that `[list("a") = list("a", "b", "a")]` being `[true, true, true]` is not intuitive, but I am not sure what would be a better, fully general behavior for `=` that's consistent with the rest of Mavo and doesn't give this result.

**Comment by LeaVerou** [comment]
I'm actually not 100% sure that `3 < 6 < undefined` being true is the most intuitive result either. Although I do see how `[1, 2, 3] < [7, 8]` might be expected to return `[true, true, true]` but I wouldn't say that's the only reasonable result, `[true, true, false]` seems reasonable as well.
`"a" = undefined` returning true however seems most certainly false.
@karger, I would appreciate some thoughts here!

**Comment by karger** [comment]
I certainly agree that "a" != undefined!
I'm skeptical that count(list("a")=list("a","b","c"))=3.  To me the correct answer is 1.   Because undefined is only equal to itself.  Although you could make a case that two undefined shouldn't be considered equal to each other either, ie undefined=undefined should return false.

I might get comfortable with the idea of right and left identities for comparison.  e.g., since 1 <4 < 7< infinity, it kind of works to call infinite the right identity for <, since you can tack it onto the end without changing the answer.  But note that this is not an identity in the sense addition because, unlike addition, the output of the comparison operator is a different type than the input.  So the notation 1<4<7<infinity is really completely different from 1+4+7+infininty; with < you are simply compressing a set of binary comparison as opposed to chaining together an associative operator.  

But my instinct says that 1<undefined should return undefined, just like 1+NaN returns NaN.

In summary, I'd never say a=undefined.  I would personally lean to 1<undefined=undefined but could accept 1<undefined=true if you are really committed to allowing unequal length

**Comment by LeaVerou** [comment]
@karger, thanks for taking a look.
As I discussed above, it seems obvious to me that `["a"] = ["a", "b", "a"]` should be `[true, false, false]` or that `"a" = undefined` should be false. What I was unsure about is what should `[4, 5, 6] < [7, 8]` return. If we use `Infinity` as a right _default_ (your comment about identities not making sense for comparison operator since their output type is different is why we renamed identity to default) then it should return `[true, true, true]` whereas if comparisons with undefined return undefined, then it should return `[true, true, undefined]` which will be treated the same as `[true, true, false]`.

Another point: `!=` is also a comparison operator, but it seems obvious to me that `a != undefined` should return true, and `["a"] != ["a", "b", "a"]` should be `[false, true, true]`. If however ANY comparison with undefined returns undefined, then this would return `[false, undefined, undefined]` which also seems unexpected.

**Comment by karger** [comment]
In your example, I currently lean to (true, false, undefined).  As you say, a lazy user can go ahead and ignore the difference, treating undefined as false, but it also permits a non-lazy user to do something different with undefined.  Forcing it to true or false as a return value would make it harder for the non-lazy user to distinguish.  

I agree that != undefined should return true in order to preserve the complementarity of = and != (it would be odd for both to be false at the same time).    There does however remain the question of weather undefined=undefined (won't come up in your example, but still).

Like you I suspect each operator needs to be considered separately.  Although, I'm relatively sanguine about returning undefined when any comparison involves undefined.  Again, this is what happens with NaN in most numeric operations, and it seems to be fine because you aren't supposed to use those results.  It's at the next level up, where for example you count(a=b), where you need to be careful to think about what to do with the undefineds your are counting.  But treating them as false seems reasonably traditional.

**Comment by LeaVerou** [comment]
@karger my main question was wrt to what should `[4,5,6] < [7,8]` return. I'm not particularly concerned about undefined vs false, since the end result in most cases is the same.

**Comment by karger** [comment]
i think it should return [true,true,undefined]

------ Original Message ------

**Comment by LeaVerou** [comment]
Ok, that makes sense. @betaveros, any chance we can get this fix into v0.2.0? 🙏
